components/scheduler2/generator/executors.go

```diff
@@ -0,0 +1,289 @@
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	TaskExecutorOptions struct {
+		fx.In
+
+		Config         *common.Config
+		MetricsHandler metrics.Handler
+		Logger         log.Logger
+		SpecBuilder    *scheduler1.SpecBuilder
+	}
+
+	taskExecutor struct {
+		TaskExecutorOptions
+	}
+)
+
+func RegisterExecutor(registry *hsm.Registry, options TaskExecutorOptions) error {
+	e := taskExecutor{options}
+	if err := hsm.RegisterTimerExecutor(registry, e.executeSleepTask); err != nil {
+		return err
+	}
+	if err := hsm.RegisterImmediateExecutor(registry, e.executeBufferTask); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e taskExecutor) executeSleepTask(env hsm.Environment, node *hsm.Node, task SleepTask) error {
+	return hsm.MachineTransition(node, func(g Generator) (hsm.TransitionOutput, error) {
+		return TransitionBuffer.Apply(g, EventBuffer{
+			Node: node,
+		})
+	})
+}
+
+func (e taskExecutor) executeBufferTask(ctx context.Context, env hsm.Environment, ref hsm.Ref, task BufferTask) error {
+	scheduler, err := common.LoadSchedulerFromParent(ctx, env, ref)
+	if err != nil {
+		return err
+	}
+	tweakables := e.Config.Tweakables(scheduler.Namespace)
+
+	generator, err := e.loadGenerator(ctx, env, ref)
+	if err != nil {
+		return err
+	}
+
+	// if we have no last processed time, this is a new schedule
+	if generator.LastProcessedTime == nil {
+		generator.LastProcessedTime = timestamppb.Now()
+		// TODO - update schedule info with create time
+
+		e.logSchedule("Starting schedule", scheduler)
+	}
+
+	// process time range between last high water mark and system time
+	t1 := timestamp.TimeValue(generator.LastProcessedTime)
+	t2 := time.Now()
+	if t2.Before(t1) {
+		e.Logger.Warn("Time went backwards",
+			tag.NewStringTag("time", t1.String()),
+			tag.NewStringTag("time", t2.String()))
+		t2 = t1
+	}
+
+	res, err := e.processTimeRange(scheduler, tweakables, t1, t2, nil)
+	if err != nil {
+		// An error here should be impossible and go to the DLQ
+		e.Logger.Error("Error processing time range", tag.Error(err))
+
+		return fmt.Errorf(
+			"%w: %w",
+			err,
+			serviceerror.NewInternal("Scheduler's Generator failed to process a time range"),
+		)
+	}
+	generator.LastProcessedTime = timestamppb.New(res.LastActionTime)
+	generator.NextInvocationTime = timestamppb.New(res.NextWakeupTime)
+
+	return env.Access(ctx, ref, hsm.AccessWrite, func(generatorNode *hsm.Node) error {
+		// Check for scheduler version conflict. We don't need to bump it, as the
+		// Generator's buffered actions are idempotent when pushed to the Executor, and the
+		// Generator's own state is also versioned. This helps Generator quickly react to
+		// schedule updates/pauses.
+		refreshedScheduler, err := hsm.MachineData[core.Scheduler](generatorNode.Parent)
+		if err != nil {
+			return err
+		}
+		if refreshedScheduler.ConflictToken != scheduler.ConflictToken {
+			// schedule was updated while processing, retry
```

Comment:
Do we actually want to retry the task here? I commented on the transition definition too but I think you can do all of this under a write lock since there's no IO involved AFAICT.

Reply:
Maybe not - If we were doing this under the write lock on the generator node, is the expectation that we'd also attempt to acquire the generator node for write access during an update request? I've been assuming that `env.Access` scopes applied only to the given `ref`, do they/can they apply more broadly?

Reply:
The lock is held for the entire mutable state record including the whole tree structure.